### PR TITLE
Fix shadow variables

### LINKED ISSUE #2491: Resolve Warnings for `-Wshadow`
| | |
|:---|:---|
|**_F´ Version_**| |
|**_Affected Component_**|  |
---
## Feature Description

Many bits of code in F´ produce warnings when compiling with `-Wshadow`.  We should fix these warnings.  The style guidelines inform us that recommended practice is to use `m_` prefixes for members to avoid this collision.

## Rationale

Code quality.

### PR DESCRIPTION
**Pull request by JohanBertrand**
| | |
|:---|:---|
|**_Related Issue(s)_**| None |
|**_Has Unit Tests (y/n)_**| N/A |
|**_Documentation Included (y/n)_**| N/A |

---
## Change Description

Add support of `-wShadow` in the code base.

Forseen changes are:
- Add a prefix `m_` to `private` and `protected` class members
- Add a prefix `a_` to arguments shadowing a `public` class member
- Change variable names to avoid shadowing class member
- Remove redundant variable definition with same name

The `a_` prefix has been chosen to avoid breaking changes in the code. Changing the name of public variables can have a bigger impact on other projects. However, I think some of those variables should become `protected` instead of `public`, and the `_m` prefix should be used on those class members in the future. I think it's an open point of discussion.

A full compatibility with `-wShadow` will require some changes also in fpp. See https://github.com/fprime-community/fpp/pull/381
I am not sure what would be the release schedule. We might want to keep the `-Wshadow` flag out until the next fpp release happens.

## Rationale

Shadow variables can make the code more difficult to read and debug.

## Testing/Review Recommendations

TBD

### CODE REVIEW

File: CMakeLists.txt
Comment (LeStarch) [inline CMakeLists.txt]: If I find any other needed changes, I'd recommend adding a `TODO` or a `FIXME` so that it is obvious that a future change is necessary.

If no other changes are needed, this is non sufficient to hold up the merge.

File: Fw/FilePacket/DataPacket.cpp
Comment (LeStarch) [inline Fw/FilePacket/DataPacket.cpp:21]: We should make two issues for this PR:
1. Create an issue to track the removal of shadows, so this work ties to an issue
2. Create an issue to track the future work (e.g. reviewing and removing `a_` shadow collisions with `public` members.

File: Ref/CMakeLists.txt
Comment (LeStarch) [inline Ref/CMakeLists.txt:60]: This should remain commented out until we fix the autocoder.  Otherwise `Ref` won't build warning-free.

File: STest/STest/Rule/Rule.hpp
Comment (LeStarch) [inline STest/STest/Rule/Rule.hpp:30]: @bocchino we will need to look into this w.r.t. the STest repository.  If and when we move to a full sub-repo, we'll need to ensure shadows are weeded-out.

I'll leave that to you to track before we make STest a subrepo.

File: Svc/FileUplink/FileUplink.hpp
Comment (LeStarch) [inline Svc/FileUplink/FileUplink.hpp:181]: Another minor nit, not blocking for this PR....but given this PR is changing much, could we rename this to `m_warning_count` or something more descriptive?

**Review by LeStarch — CHANGES_REQUESTED** [review]
I think there is only one required change I see: comment out the `-Wshadow` in the 
Ref package until the PR on FPP makes its way over.

If you are so-motivated, there are a few other minor comments on other cleanup items....however, none of these are required in this PR.  I'll take the action to add tickets as we need something to track this work.

@bocchino or @thomas-bc this is a big PR but entirely formulaic, would one of you take a look to ensure I did not miss anything?

**Comment by LeStarch** [comment]
The CI failures are likely due to unit testing.  Much of our unit testing relies on white-box testing and as such does access these private members.  We'll need to fix those too.  @JohanBertrand are you up for this fix?

File: Fw/FilePacket/DataPacket.cpp
Comment (LeStarch) [reply Fw/FilePacket/DataPacket.cpp:21]: I made both issues.  Hope I am not stepping on toes.

**Comment by LeStarch** [comment]
In case this isn't clear, this is solid work.  Thanks for the contribution.  Feel free to communicate us if you have reached the limit of what you are able to work on and we can move it through the process, otherwise we'll let you move it forward!

**Comment by JohanBertrand** [comment]
> I think there is only one required change I see: comment out the `-Wshadow` in the Ref package until the PR on FPP makes its way over.

From what I've seen, it works well because the warning flags apply only to the user code, and not to the auto generated code.  The unit tests and the build on Ref were working fine for me.

> If you are so-motivated, there are a few other minor comments on other cleanup items....however, none of these are required in this PR. I'll take the action to add tickets as we need something to track this work.

I'm happy keeping the work going, all good. I would definitely do the fix of the `a_` prefix in another PR to make it easier to review, but I'll go through the rest.

> The CI failures are likely due to unit testing. Much of our unit testing relies on white-box testing and as such does access these private members. We'll need to fix those too. @JohanBertrand are you up for this fix?

It was compiling well for me. I already did changes in the test files, but just a few were needed. I did not run the `check` though. I'll double check, thanks.

Does this relate to the failure in `CI / Framework (pull_request)`?

> In case this isn't clear, this is solid work. Thanks for the contribution. Feel free to communicate us if you have reached the limit of what you are able to work on and we can move it through the process, otherwise we'll let you move it forward!

Thanks for the kind words, I'll come back with the changes in a few hours/days.

File: Ref/CMakeLists.txt
Comment (JohanBertrand) [reply Ref/CMakeLists.txt:60]: Correct me if I am wrong, but it looks like the warning here only affects the code written in the Ref project, but does not affect the code generated by fpp

File: Svc/FileUplink/FileUplink.hpp
Comment (JohanBertrand) [reply Svc/FileUplink/FileUplink.hpp:181]: I changed the `m_n` variable names to something more descriptive, and updated the comments of the variables. Let me know if there is a mistake in the comments

**Comment by JohanBertrand** [comment]
@LeStarch it should be good to go, let me know if something else is need

I'll work on https://github.com/nasa/fprime/issues/2492 once this PR is closed

File: CMakeLists.txt
Comment (JohanBertrand) [reply CMakeLists.txt]: Added a TODO

**Comment by LeStarch** [comment]
I fixed a comment.  We'll set what CI does, if it fails I can render more advice!

**Comment by JohanBertrand** [comment]
Is there anything else needed before merging?

File: Ref/CMakeLists.txt
Comment (LeStarch) [reply Ref/CMakeLists.txt:60]: True, this only affects `Ref`

**Review by LeStarch — APPROVED** [review]

**Comment by LeStarch** [comment]
I have approved, was hoping for one more set of eyes on this because of the extensive length of the PR.  Once we get that it will be merged!

**Review by thomas-bc — APPROVED** [review]

**Pull request merged**
